### PR TITLE
fix: fix conversion for pre-epoch timestamps

### LINKED ISSUE #159: Timestamp.of(java.sql.Timestamp) off-by-one on pre-epoch dates
Originally reported by @s13o in https://github.com/spring-cloud/spring-cloud-gcp/issues/2041

### PR DESCRIPTION
**Pull request by chingor13**
Fixes #159

### CODE REVIEW

File: google-cloud-core/src/test/java/com/google/cloud/TimestampTest.java [hunk 1]
```diff
@@ -80,6 +80,30 @@ public void ofDate() {
     assertThat(timestamp.getNanos()).isEqualTo(expectedNanos);
   }
 
+  @Test
+  public void ofSqlTimestamp() {
+    String expectedTimestamp = "1970-01-01T00:00:12.345000000Z";
```
Comment (elharo) [inline google-cloud-core/src/test/java/com/google/cloud/TimestampTest.java]: This is a string, not a timestamp

File: google-cloud-core/src/main/java/com/google/cloud/Timestamp.java [hunk 2]
```diff
@@ -119,7 +119,13 @@ public static Timestamp now() {
    * @throws IllegalArgumentException if the timestamp is outside the representable range
    */
   public static Timestamp of(java.sql.Timestamp timestamp) {
-    return ofTimeSecondsAndNanos(timestamp.getTime() / 1000, timestamp.getNanos());
+    // TODO: replace with Math.floorDiv when we drop Java 7 support
+    long secs = timestamp.getTime() / 1000;
```
Comment (elharo) [inline google-cloud-core/src/main/java/com/google/cloud/Timestamp.java]: secs --> seconds per Google style

**Review by codyoss — APPROVED** [review]

File: google-cloud-core/src/main/java/com/google/cloud/Timestamp.java [hunk 3]
```diff
@@ -119,7 +119,13 @@ public static Timestamp now() {
...
+    int nanos = timestamp.getNanos();
+    if (secs < 0) {
+      --secs;
```
Comment (elharo) [inline google-cloud-core/src/main/java/com/google/cloud/Timestamp.java]: This feels like a hack. Why does this happen in the first place? Is this expected behavior?

Either way, it should probably have comments in the code.

**Review by elharo — COMMENTED** [review]
also, haven't we seen this problem before a few months back? maybe in a different project?

File: google-cloud-core/src/main/java/com/google/cloud/Timestamp.java (hunk 3 above)
Comment (chingor13) [reply google-cloud-core/src/main/java/com/google/cloud/Timestamp.java]: Added a comment about why the hack is in there and a TODO for switching to `Math.floorDiv` when we drop Java 7 support

**Comment by chingor13** [comment]
> also, haven't we seen this problem before a few months back? maybe in a different project?

I think this was something similar: https://github.com/googleapis/google-http-java-client/pull/895

**Pull request merged**